### DISCUSSION TITLE
How to build Risc-V architecture UEFI

### DISCUSSION THREAD

**Discussion by wjh-study** [tianocore/edk2#5040] [Q&A]
Now I download the edk2 and edk2-platforms then I git submodule update --init successfully.
But,I want to build Risc-V architecture EFI that I can run in the Risc-V FPGA.
So,I want to know that what libraries are necessary for me such as PEI, DXE, opensbi.
I referred to Sifive U500 but it has too many library that i do not need.
I want to a clean UEFI that can boot in FPGA.May I ask if anyone can provide assistance

**Comment by vlsunil** [comment]
Hi @wjh-study : You can refer to https://github.com/tianocore/edk2-platforms/tree/master/Platform/Sophgo/SG2042_EVB_Board. There is another platform whose support is in-flight and not yet merged which needs https://edk2.groups.io/g/devel/message/110778 and https://edk2.groups.io/g/devel/message/110589. You can also look at https://github.com/tianocore/edk2/tree/master/OvmfPkg/RiscVVirt for reference virtual platform.

**Reply by wjh-study** [reply]
When I follow the virtual platform,It need openSUSE to start RiscV virt.But my machine is Ubuntu.
I am a green hand in edk2.But I have compiled the opensbi and edk2(RiscV Virt).how can I start in Ubuntu?
Now I have RISCV_VIRT_CODE.fd, RISCV_VIRT_VARS.fd and fw_payload.bin.But I can not start RiscV virt with Qemu.
May I ask if anyone can provide assistance

**Reply by vlsunil** [reply]
As it is mentioned, OpenSuse boot was just an example. You can boot Ubuntu 23.10 also by using preinstalled image for Qemu Emulator (https://ubuntu.com/download/risc-v). You  can also use direct kernel boot.

**Reply by wjh-study** [reply]
Hi vlsunil:I want to try run uefi with qemu..I have try my best to find a cmd to start qemu with the RISCV_VIRT_CODE.fd, RISCV_VIRT_VARS.fd and fw_payload.bin,but I can not find.....I want to how to use RISCV_VIRT_CODE.fd, RISCV_VIRT_VARS.fd and fw_payload.bin to start riscv qemu to run uefi.....
I want to run my code in edk2 with qemu..and I have compiled my code..How can I run the uefi with qemu?.I want to how to use RISCV_VIRT_CODE.fd, RISCV_VIRT_VARS.fd and fw_payload.bin to start riscv qemu to run uefi......
I may not have expressed myself clearly before.Thanks very much

**Reply by vlsunil** [reply]
I am not sure I understand what help you need. https://github.com/tianocore/edk2/blob/master/OvmfPkg/RiscVVirt/README.md#test already documents the qemu command line about how to boot qemu virt machine with RISCV_VIRT_CODE.fd and RISCV_VIRT_VARS.fd. If you want to build OpenSBI yourself instead of using the default one that comes with qemu build, you have to use fw_dynamic.bin for -bios as mentioned in https://github.com/tianocore/edk2/blob/master/OvmfPkg/RiscVVirt/README.md#test-with-your-own-opensbi-binary.

**Reply by wjh-study** [reply]
Hi vlsunil:I used cmd 
qemu-system-riscv64 -M virt -m 4096 -blockdev node-name=pflash0,driver=file,read-only=on,filename=RISCV_VIRT_CODE.fd -blockdev node-name=pflash1,driver=file,filename=RISCV_VIRT_VARS.fd -serial mon:stdio -device qemu-xhci -bios /mnt/disk/wjh/code/edk/edk2-platforms/Silicon/RISC-V/ProcessorPkg/Library/RiscVOpensbiLib/opensbi/build/platform/generic/firmware/fw_dynamic.bin.
 I saw the SecMainRiscV64 (Fixed Flash Address, BaseAddress=0x00200000c0, EntryPoint=0x0020000300, Type=PE).
so I set opensbi next_addr is 0x20000300.But I still can not start uefi....shutdown in opensbi.

**Reply by vlsunil** [reply]
Hi, qemu options internally set the required addresses. So, please use the command line exactly as is for both build and test.

**Reply by vlsunil** [reply]
Please build opensbi from https://github.com/riscv-software-src/opensbi

**Reply by vlsunil** [reply]
What is the qemu version you are using? It should be 8.1 at the minimum as mentioned in the readme.

**Reply by wjh-study** [reply]
Hi vlsunil:I am very grateful for your help and I hope you have a very pleasant weekend.
Now,the qemu I used is "QEMU emulator version 8.1.90", and when I run the cmd, it shows "device requires 33554432 bytes, block backend provides 8388608 bytes" but I set the memory is "-m 40960"

**Reply by vlsunil** [reply]
Please follow step 1 in https://github.com/tianocore/edk2/tree/master/OvmfPkg/RiscVVirt#test.

**Reply by wjh-study** [reply]
Hi: Thanks for your help，I have run the edk2 with qemu successfully.
